CMakeLists.txt

```diff
@@ -0,0 +1,129 @@
+cmake_minimum_required(VERSION 3.2)
+project(libxmp VERSION 4.5.0 LANGUAGES C)
+
+include(build_pros.cmake)
+
+# -fPIC thing
+if(NOT WIN32 AND NOT EMSCRIPTEN AND NOT VITA)
+    option(LIBXMP_PIC   "Build the static library as the position-independent" OFF)
+else()
+    set(LIBXMP_PIC OFF)
+endif()
+# -----------------------------------
+
+if(EMSCRIPTEN OR VITA)
+    set(BUILD_SHARED_DEFAULT OFF)
+else()
+    set(BUILD_SHARED_DEFAULT ON)
+endif()
+
+option(BUILD_STATIC   "Build static library of libXMP" ON)
+option(BUILD_SHARED   "Build shared library of libXMP" ${BUILD_SHARED_DEFAULT})
+if(NOT BUILD_STATIC AND NOT BUILD_SHARED)
+    message(FATAL_ERROR "Both static and shared builds got disabled. You must enable at least one of them.")
+endif()
+
+option(LIBXMP_DISABLE_DEPACKERS     "Disable archive depackers at libXMP" OFF)
+option(LIBXMP_DISABLE_PROWIZARD     "Disable the ProWizard at libXMP" OFF)
+
+include(build_src.cmake)
+
+
+set(LIBXMP_DEFINES)
+
+if(LIBXMP_DISABLE_DEPACKERS)
+    list(APPEND LIBXMP_DEFINES -DLIBXMP_NO_DEPACKERS)
+else()
+    list(APPEND LIBXMP_SRC_LIST_FULL ${LIBXMP_SRC_LIST_DEPACKERS})
+endif()
+
+if(LIBXMP_DISABLE_PROWIZARD)
+    list(APPEND LIBXMP_DEFINES -DLIBXMP_NO_PROWIZARD)
+else()
+    list(APPEND LIBXMP_SRC_LIST_FULL ${LIBXMP_SRC_LIST_PROWIZARD})
+endif()
+
+list(APPEND LIBXMP_SRC_FULL ${LIBXMP_SRC_LIST_LITE} ${LIBXMP_SRC_LIST_FULL} ${LIBXMP_SRC_COMMON})
+
+if(MSVC)
+    list(APPEND LIBXMP_DEFINES -DDinline=__inline -D_USE_MATH_DEFINES -DPATH_MAX=260)
+endif()
+
+set(XMP_INSTALLS)
+
+# ============= FULL =============
+if(BUILD_STATIC)
+    add_library(xmp_static STATIC ${LIBXMP_SRC_FULL})
+    list(APPEND XMP_INSTALLS xmp_static)
+    set_target_properties(xmp_static PROPERTIES OUTPUT_NAME xmp C_STANDARD 90)
+
+    if(WIN32 AND BUILD_SHARED)
+        set_target_properties(xmp_static PROPERTIES OUTPUT_NAME xmp-static)
+    else()
+        set_target_properties(xmp_static PROPERTIES OUTPUT_NAME xmp)
+    endif()
+
+    target_compile_definitions(xmp_static PRIVATE ${LIBXMP_DEFINES} -DBUILDING_STATIC)
```

Comment:
With the exception of the MSVC compatibility defines, it would be good to make these defines public.

Reply:
I see

BTW, one small note: I saw the thing at the `xmp.h` that there were "dllimport()" markers that were been used. I usually do NOT use them, mainly because I want to allow easy use of the library with both static and shared form without the necessity to define extra macros at the user side, otherwise, when "dllimport()" was been used at the static library, the link will fail.

So, the little patch example I made at the my ModPlug copy:
https://github.com/WohlSoft/AudioCodecs/commit/042b59701adba4f6201f867e006a1f887367fcb2
So, never use `dllimport()` at libraries intended to be both static and shared to simplify the user's choice of which library to link. Otherwise, this adds an extra pain to developers that require them to specify extra macros to link the same library statically.

Reply:
I heard that `dllimport()` helps to initialize the shared library faster, however, I do see not so much difference in the workflow when using the same application with or without the `dllimport()`. But, I do think, when the library is intended to be both shared and static, so, don't use `dllimport()` at all, otherwise this adds an extra pain for users, letting them figure why your static library won't link to their applications.

Reply:
btw, I guess I should only make the `-DBUILDING_STATIC` macro defined, there are other macros that appeared here, and the `-DBUILDING_DLL` macro that appears at the shared library. So, I guess, I should don't publicize other macros to not confuse the whole thing, the `-DBUILDING_STATIC` will be published.

Reply:
Or, what do you think? @ccawley2011

Reply:
> [...] `dllimport()` [...]

Off-topic

Reply:
> btw, I guess I should only make the `-DBUILDING_STATIC` macro defined, there are other macros that appeared here, and the `-DBUILDING_DLL` macro that appears at the shared library. So, I guess, I should don't publicize other macros to not confuse the whole thing, the `-DBUILDING_STATIC` will be published.

IMO,  `-DBUILDING_STATIC` and `-DBUILDING_DLL` should be private to build of libxmp.
If some other library is explicitly building against static libxmp, it should define `BUILDING_STATIC`
in its relevant compilation unit for that purpose. 

Otherwise, unforeseen complications might occur (because these are unfortunately fairly
common macro names.)

Reply:
> Off-topic

moved: https://github.com/libxmp/libxmp/issues/427

Reply:
It would be worth making `LIBXMP_NO_DEPACKERS`, `LIBXMP_NO_PROWIZARD` and `LIBXMP_CORE_DISABLE_IT` public, since all three defines are prefixed with "LIBXMP_", and it avoids the need to manually add them again when building the tests.

Reply:
Sounds reasonable, will do that.